### PR TITLE
SAP File Server MD

### PR DESCRIPTION
**Pull request by MSSedusch**
### Best Practice Checklist
Check these items before submitting a PR... See the Contribution Guide for the full detail: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/README.md 

1. uri's compatible with all clouds (Stack, China, Government)
1. Staged artifacts use _artifactsLocation & _artifactsLocationSasToken
1. Use resourceGroup().location for resource locations
1. Folder names for artifacts (nestedtemplates, scripts, DSC)
1. Use literal values for apiVersion (no variables)
1. Parameter files (GEN-UNIQUE for value generation and no "changemeplease" values
1. $schema and other uris use https
1. Use uniqueString() whenever possible to generate names for resources.  While this is not required, it's one of the most common failure points in a deployment. 
1. Update the metadata.json with the current date

For details: https://github.com/Azure/azure-quickstart-templates/blob/master/1-CONTRIBUTION-GUIDE/bp-checklist.md

- [x] - Please check this box once you've submitted the PR if you've read through the Contribution Guide and best practices checklist.

### Changelog

* add new file server template for use with SAP NetWeaver

### CODE REVIEW

**Comment by akhil-thomas011** [comment]
Can you please do a dummy commit again?

**Comment by MSSedusch** [comment]
@akhil-thomas011 done. checks passed

File: sap-file-server-md/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,493 @@
...
+    "subnetId": {
+      "type": "string",
+      "defaultValue": "",
+      "metadata": {
+        "description": "The id of the subnet you want to use."
+      }
+    }
+  },
+  "variables": {
+    "github": "https://raw.githubusercontent.com/Azure/azure-quickstart-templates/master/sap-2-tier-marketplace-image/shared/",
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: see BP guide for how to handle artifacts

File: sap-file-server-md/azuredeploy.parameters.json [hunk 2]
```diff
@@ -0,0 +1,15 @@
+{
+  "$schema": "http://schema.management.azure.com/schemas/2015-01-01/deploymentParameters.json#",
+  "contentVersion": "1.0.0.0",
+  "parameters": {
+    "resourcePrefix": {
+      "value": "DEV"
+    },
+    "adminUsername": {
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.parameters.json]: GEN-UINIQUE

File: sap-file-server-md/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+      "SLES 12 BYOS": {
+        "sku": "12-SP1",
+        "offer": "SLES-SAP-BYOS",
+        "publisher": "SUSE",
+        "OSType": "Linux",
+        "version": "latest",
+        "Plan": {},
+        "UsePlan": false
+      },
+      "RHEL 7.2": {
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: is this ever used?

File: sap-file-server-md/azuredeploy.json [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+        "sku": "7.2",
+        "offer": "RHEL",
+        "publisher": "RedHat",
+        "version": "latest",
+        "OSType": "Linux",
+        "Plan": {},
+        "UsePlan": false
+      },
+      "Oracle Linux 7.2": {
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: is this ever used

File: sap-file-server-md/azuredeploy.json [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+        "sku": "7.2",
+        "offer": "Oracle-Linux",
+        "publisher": "Oracle",
+        "version": "latest",
+        "OSType": "Linux",
+        "Plan": {},
+        "UsePlan": false
+      }
+    },
+    "imageSku": "[variables('images')[parameters('osType')].sku]",
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: A lot of these variables are only used once just for wrapping the object - removing will simplify the template quite a bit.

File: sap-file-server-md/azuredeploy.json [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+    "nfsvmDataDisks": "[variables('vmSizes').nfsserverdisks]",
+    "sidlower": "[toLower(parameters('resourcePrefix'))]",
+    "vmName": "[variables('sidlower')]",
+    "vnetName": "[concat(variables('sidlower'), '-vnet')]",
+    "publicIPAddressType": "Dynamic",
+    "addressPrefix": "10.0.0.0/16",
+    "subnetName": "Subnet",
+    "subnetPrefix": "10.0.0.0/24",
+    "subnets": {
+      "new": "[concat(resourceId('Microsoft.Network/virtualNetworks', variables('vnetName')), '/subnets/', variables('subnetName'))]",
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: just use resourceId, remove concat

File: sap-file-server-md/azuredeploy.json [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+  },
+  "resources": [
+    {
+      "apiVersion": "2015-01-01",
+      "name": "[variables('nestedDeploymentNameNSG')]",
+      "type": "Microsoft.Resources/deployments",
+      "properties": {
+        "mode": "Incremental",
+        "templateLink": {
+          "uri": "[concat(variables('github'), parameters('newOrExistingSubnet'), 'nsg.json')]",
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: this is an old pattern - use the condition property on the resources instead - same for the others below

File: sap-file-server-md/azuredeploy.json [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+      "name": "[concat(variables('nestedDeploymentNamePIP'), '-', copyIndex())]",
+      "type": "Microsoft.Resources/deployments",
+      "copy": {
+        "name": "pipLoop",
+        "count": "[variables('nfsvmCount')]"
+      },
+      "properties": {
+        "mode": "Incremental",
+        "templateLink": {
+          "uri": "[concat(variables('github'), parameters('newOrExistingSubnet'), 'pip.json')]",
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: why is this conditional and the nic it's assigned to is not?

File: sap-file-server-md/azuredeploy.json [hunk 9]
```diff
@@ -0,0 +1,493 @@
...
+      }
+    },
+    "adminPassword": {
+      "type": "securestring",
+      "minLength": 12,
+      "metadata": {
+        "description": "Password for the Virtual Machine."
+      }
+    },
+    "newOrExistingSubnet": {
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: remove this (see below)

File: sap-file-server-md/azuredeploy.json [hunk 10]
```diff
@@ -0,0 +1,493 @@
...
+      "defaultValue": "new",
+      "allowedValues": [
+        "new",
+        "existing"
+      ],
+      "metadata": {
+        "description": "Determines whether a new virtual network and subnet should be created or an existing subnet should be used."
+      }
+    },
+    "subnetId": {
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: use the condition property to determine when to deploy the resource - if this is empty, deploy the network - and the you can update the description for instruction... e.g. enter it subnetId to use, if blank a new virtualNetwork will be provisioned.

File: sap-file-server-md/azuredeploy.parameters.json (hunk 2 above)
Comment (MSSedusch) [inline sap-file-server-md/azuredeploy.parameters.json]: fixed

File: sap-file-server-md/azuredeploy.json [hunk 11]
```diff
@@ -0,0 +1,449 @@
...
+      "type": "string",
+      "defaultValue": "",
+      "metadata": {
+        "description": "The id of the subnet you want to use."
+      }
+    },
+    "_artifactsLocation": {
```
Comment (bmoore-msft) [inline sap-file-server-md/azuredeploy.json]: need a param for sasToken too

File: sap-2-tier-marketplace-image/shared/nic-config.json [hunk 12]
```diff
@@ -31,12 +31,12 @@
       }
     },
     "newOrExistingSubnet": {
```
Comment (bmoore-msft) [inline sap-2-tier-marketplace-image/shared/nic-config.json]: This still seems like an odd pattern to me but I think we're stuck with it until we get if/then support - a lot of misdirection :)

File: sap-2-tier-marketplace-image/shared/loadbalancer.json [hunk 13]
```diff
@@ -60,45 +60,63 @@
...
-            }
+            },
+            "defaultValue": ""
+        },
+        "_artifactsLocation": {
+            "type": "string",
+            "metadata": {
+                "description": "The base URI where artifacts required by this template are located. When the template is deployed using the accompanying scripts, a private location in the subscription will be used and this value will be automatically generated."
+            },
+            "defaultValue": ""
```
Comment (bmoore-msft) [inline sap-2-tier-marketplace-image/shared/loadbalancer.json]: will need default value for when CI is fixed

File: sap-2-tier-marketplace-image/shared/loadbalancer.json [hunk 14]
```diff
@@ -60,45 +60,63 @@
...
         {
-            "apiVersion": "[variables('apiVersionDeployment')]",
+            "apiVersion": "2017-05-10",
             "name": "[concat(variables('nestedDeploymentNameLBSCS'), '0')]",
             "type": "Microsoft.Resources/deployments",
             "properties": {
                 "mode": "Incremental",
                 "templateLink": {
-                    "uri": "[concat(parameters('github'), 'lb-SCS0.json')]",
+                    "uri": "[concat(variables('selectedArtifacts'), 'lb-SCS0.json', parameters('_artifactsLocationSasToken'))]",
```
Comment (bmoore-msft) [inline sap-2-tier-marketplace-image/shared/loadbalancer.json]: need '/' separator before the filename

File: sap-file-server-md/metadata.json [hunk 15]
```diff
@@ -0,0 +1,7 @@
+{
+  "itemDisplayName": "SAP NetWeaver file server - MD",
```
Comment (bmoore-msft) [inline sap-file-server-md/metadata.json]: MD might be ambiguous to some folks - suggest: "SAP NetWeaver File Server (Managed Disks)"

File: sap-file-server-md/shared/lb-DB-HANA.json [hunk 16]
```diff
@@ -0,0 +1,165 @@
...
+        "probesArray": {
+            "type": "array",
+            "defaultValue": [],
+            "metadata": {
+                "description": "The array that contains the probes for the Load Balancer creation"
+            }
+        }
+    },
+    "variables": {
+        "apiVerion": "2015-06-15",
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/lb-DB-HANA.json]: remove - not used

File: sap-file-server-md/shared/lb-DB-SQL.json [hunk 17]
```diff
@@ -0,0 +1,146 @@
...
+        "probesArray": {
+            "type": "array",
+            "defaultValue": [],
+            "metadata": {
+                "description": "The array that contains the probes for the Load Balancer creation"
+            }
+        }
+    },
+    "variables": {
+        "apiVerion": "2015-06-15",
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/lb-DB-SQL.json]: remove

File: sap-file-server-md/shared/lb-ERS-Linux.json [hunk 18]
```diff
@@ -0,0 +1,301 @@
...
+        "probesArray": {
+            "type": "array",
+            "defaultValue": [],
+            "metadata": {
+                "description": "The array that contains the probes for the Load Balancer creation"
+            }
+        }
+    },
+    "variables": {
+        "apiVerion": "2015-06-15",
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/lb-ERS-Linux.json]: remove

File: sap-file-server-md/shared/lb-SCS.json [hunk 19]
```diff
@@ -0,0 +1,463 @@
...
+        "probesArray": {
+            "type": "array",
+            "defaultValue": [],
+            "metadata": {
+                "description": "The array that contains the probes for the Load Balancer creation"
+            }
+        }
+    },
+    "variables": {
+        "apiVerion": "2015-06-15",
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/lb-SCS.json]: remove

File: sap-file-server-md/shared/loadbalancer.json [hunk 20]
```diff
@@ -0,0 +1,361 @@
...
+            "false": "[concat('lb-ERS0.json')]"
+        },
+        "dbTemplates": {
+            "true": "[concat('lb-DB-', parameters('dbType'), '.json')]",
+            "false": "[concat('lb-DB0.json')]"
+        },
+        "scsTemplate": "[variables('scsTemplates')[string(parameters('createXSCS'))]]",
+        "ersTemplate": "[variables('ersTemplates')[string(parameters('createXSCS'))]]",
+        "dbTemplate": "[variables('dbTemplates')[string(parameters('createDB'))]]",
+        "artifacts": {
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/loadbalancer.json]: same comment as the other...

File: sap-file-server-md/shared/newpip.json [hunk 21]
```diff
@@ -0,0 +1,33 @@
...
+  },
+  "variables": {
+    "apiVerion": "2015-06-15",
+    "apiVerionRm": "2015-01-01"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Network/publicIPAddresses",
+      "name": "[parameters('publicIpName')]",
+      "apiVersion": "[variables('apiVerion')]",
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/newpip.json]: use literal apiVersions so schema validation can be done

File: sap-file-server-md/shared/newvnetnsg.json [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+  },
+  "variables": {
+    "apiVerion": "2015-06-15",
+    "apiVerionRm": "2015-01-01"
+  },
+  "resources": [
+    {
+      "type": "Microsoft.Network/virtualNetworks",
+      "name": "[parameters('vnetName')]",
+      "apiVersion": "[variables('apiVerion')]",
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/newvnetnsg.json]: use literal apiVersions

File: sap-file-server-md/shared/noop.ps1 [hunk 23]
```diff
@@ -0,0 +1,28 @@
+﻿param
+(
+    [String] $DBDataLUNS = "0,1,2",	
+    [String] $DBLogLUNS = "3",
+    [string] $DBDataDrive = "S:",
+    [string] $DBLogDrive = "L:",
+	[string] $DBDataName = "dbdata",
```
Comment (bmoore-msft) [inline sap-file-server-md/shared/noop.ps1]: formatting

**Pull request merged**